pallet/incentive/src/lib.rs

```diff
@@ -0,0 +1,811 @@
+			let who = ensure_signed(origin)?;
+			let vault_account = Self::get_vault_account(id).unwrap();
+
+			let Some(user_info) = PoolUsers::<T>::get(id, &who) else {
+				Err(Error::<T>::NoTokensStaked)?
+			};
+			let Some(pool) = Pools::<T>::get(id) else {
+				Err(Error::<T>::PoolDoesNotExist)?
+			};
+
+			ensure!(
+				frame_system::Pallet::<T>::block_number() > pool.end_block,
+				Error::<T>::NotReadyForClaimingReward
+			);
+
+			let reward = Self::calculate_reward(&user_info, &pool);
+			if reward == Zero::zero() {
+				return Ok(())
+			}
+
+			let amount = if user_info.should_rollover == false {
+				T::Assets::transfer(pool.asset_id, &vault_account, &who, user_info.amount, false)?;
+				user_info.amount
+			} else {
+				Zero::zero()
+			};
+
+			// Transfer reward to user
+			T::Currency::transfer(&vault_account, &who, reward, ExistenceRequirement::AllowDeath)?;
+
+			Pools::<T>::try_mutate(id, |pool_info| -> DispatchResult {
+				let pool_info = pool_info.as_mut().ok_or(Error::<T>::PoolDoesNotExist)?;
+				pool_info.last_updated = frame_system::Pallet::<T>::block_number();
+				pool_info.locked_amount = pool_info.locked_amount.saturating_sub(amount);
+				Ok(())
+			})?;
+			PoolUsers::<T>::try_mutate(id, who, |pool_user| -> DispatchResult {
+				let pool_user = pool_user.as_mut().ok_or(Error::<T>::NoTokensStaked)?;
+				pool_user.reward_debt = pool_user.reward_debt.saturating_add(reward);
+				Ok(())
+			})?;
+
+			Self::deposit_event(Event::RewardsClaimed(who, id, reward));
+			Ok(())
+		}
+
+		#[pallet::weight(T::WeightInfo::rollover_unsigned())]
+		#[transactional]
+		pub fn rollover_unsigned(
+			origin: OriginFor<T>,
+			id: T::PoolId,
+			_current_block: BlockNumberFor<T>,
+		) -> DispatchResult {
+			ensure_none(origin)?;
+			let Some(pool_info) = Pools::<T>::get(id) else {
+				return Err(Error::<T>::PoolDoesNotExist.into())
+			};
+			log::warn!("start processing the rollover");
+			if let PoolStatus::RollingOver = pool_info.pool_status {
+				if let Some(vault_account) = Self::get_vault_account(id) {
+					// Check for successor
+					let successor_id =
+						PoolRelationships::<T>::get(id).unwrap_or_default().successor_id;
+
+					if let Some(successor_id) = successor_id {
+						let Some(successor_pool_info) = Pools::<T>::get(successor_id) else {
+							return Err(Error::<T>::PoolDoesNotExist.into())
+						};
+						// Migrate users to successor
+						let start_key = Self::rollover_pivot(id);
+						let mut map_iterator = match RolloverPivot::<T>::contains_key(id) {
+							true => <PoolUsers<T>>::iter_prefix_from(id, start_key.clone()),
+							false => <PoolUsers<T>>::iter_prefix(id),
+						};
+
+						let mut count = 0;
+						let mut total_rolled_over_amount = successor_pool_info.locked_amount;
+
+						let successor_vault_account =
+							Self::get_vault_account(successor_id).unwrap();
+
+						while let Some((who, user_info)) = map_iterator.next() {
+							if Self::should_migrate(&user_info) {
```

Comment:
applying suggestion above (moving the function to the struct impl)
```suggestion
							if &user_info.should_migrate() {
```